source/adapters/hip/program.cpp

```diff
@@ -74,17 +74,35 @@ void getCoMgrBuildLog(const amd_comgr_data_set_t BuildDataSet, char *BuildLog,
 } // namespace
 #endif
 
+std::pair<std::string, std::string>
+splitMetadataName(const std::string &metadataName) {
+  size_t splitPos = metadataName.rfind('@');
+  if (splitPos == std::string::npos)
+    return std::make_pair(metadataName, std::string{});
+  return std::make_pair(metadataName.substr(0, splitPos),
+                        metadataName.substr(splitPos, metadataName.length()));
+}
+
 ur_result_t
 ur_program_handle_t_::setMetadata(const ur_program_metadata_t *Metadata,
                                   size_t Length) {
   for (size_t i = 0; i < Length; ++i) {
     const ur_program_metadata_t MetadataElement = Metadata[i];
     std::string MetadataElementName{MetadataElement.pName};
 
+    auto [Prefix, Tag] = splitMetadataName(MetadataElementName);
+
     if (MetadataElementName ==
         __SYCL_UR_PROGRAM_METADATA_TAG_NEED_FINALIZATION) {
       assert(MetadataElement.type == UR_PROGRAM_METADATA_TYPE_UINT32);
       IsRelocatable = MetadataElement.value.data32;
+    } else if (Tag == __SYCL_UR_PROGRAM_METADATA_GLOBAL_ID_MAPPING) {
+      const char *MetadataValPtr =
+          reinterpret_cast<const char *>(MetadataElement.value.pData) +
+          sizeof(std::uint64_t);
```

Comment:
Why the offset (here addition and in the `MetadataValPtrEnd` subtraction)?

Reply:
This seems to just be how the metadata is laid out, have lifted this directly from the CUDA adapter, and it works so not going to dig too much into this